Fix use of 'is' operator for comparison

tambry opened the pull request:
The 'is' operator is not meant to be used for comparisons. It currently working is an implementation detail of CPython.
CPython 3.8 has added a SyntaxWarning for this.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/Vulkan-ValidationLayers?pullRequest=904) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/Vulkan-ValidationLayers?pullRequest=904) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/Vulkan-ValidationLayers?pullRequest=904) it.</sub>

mark-lunarg commented:
Hey @tambry, any idea why CI is so sad?  Seems like one or more of these changes might not be compatible with python 3.5.2 or 3.5.4 (Linux and Windows)?

mikes-lunarg commented:
@mark-lunarg it does not look like an issue with a specific python version, it breaks on my system as well.  While the changes look harmless enough, they result in an empty vk_safe_struct.h/.cpp.

tambry commented:
This seemed like an obvious change, so I didn't test it locally.

I added this print before the affected if in `helper_file_generator.py`:

    print(f"{elem.tag}, {elem.tag is not 'type'}, {elem.tag != 'type'}")

And the output looks like this:
```
name, True, True
type, True, False
name, True, True
type, True, False
[100s more ...]
```

Turns out that CPython is not actually doing object pooling in this case thus making `elem.tag is not 'type'` *always* true. Most of the cases I've fixed previously have been with `''`, which CPython does pool.

I guess this should be handed over to someone who knows what the code is meant to be doing? Simply removing the condition doesn't seem like a good idea.

mikes-lunarg commented:
I've done some more digging and it seems this change has inadvertently demonstrated why using 'is' with literals is so dangerous. As you noted, `elem.tag is not 'type'` is *always* evaluating to `True` regardless of its value.  Something about the way `elem.tag` is constructed is preventing it from using the string interning optimization and it is therefore distinct from the literal `'type'`

After looking at the xml that this code is trying to parse I believe the correct implementation is:

            if elem.tag == 'type' and elem.tail is not None and '*' in elem.tail:
                ispointer = True

It will still work without the first check, but it is more correct to ensure that the `*` character is found directly after the type tag

tambry commented on scripts/parameter_validation_generator.py:
```diff
@@ -361,10 +361,10 @@ def endFile(self):
                     pnext_case += '#endif // %s\n' % protect
                 # Skip functions containing no validation
-                if struct_validation_source != '':
+                if struct_validation_source:
```
I changed this for consistency with the if above.

tambry commented:
Thanks for the digging!

I've applied the fix you suggested.

jzulauf-lunarg commented:
@mikes-lunarg -- did you pull this for a local CI cycle?

mikes-lunarg commented:
@jzulauf-lunarg I did not run it through the local farm but tested on my machine against python 3.6, 3.7, and 3.8a3.  All 3 versions of python I ran it through produced identical codegen as before the change.
